### PR TITLE
Handle URL base

### PR DESCRIPTION
**Pull request by Zuluru**
Addresses issue #230.

### CODE REVIEW

**Comment by Zuluru** [comment]
Hmm, looks like the `base` is being incorrectly set to the `vendor` folder during unit testing. Wasn't a problem until now. Will need a similar fix to the one in the main cakephp repo, I guess.

**Comment by Zuluru** [comment]
We probably want to implement a better fix for the unit tests, like PR12076 in the cakephp repo. The fix used here is the previous, insufficient one from there, just to demonstrate that the Travis issues with the previous commit are in the unit tests rather than my changes, to allow discussion to proceed.

File: src/Middleware/AuthenticationMiddleware.php [hunk 1]
```diff
@@ -146,7 +146,9 @@ protected function getRedirectUrl($target, ServerRequestInterface $request)
             return $target;
         }
 
-        $query = urlencode($param) . '=' . urlencode($request->getUri());
+        $uri = $request->getUri();
+        $uri = $uri->withPath($uri->base . $uri->getPath());
```
Comment (markstory) [inline src/Middleware/AuthenticationMiddleware.php]: Should we make sure that the uri has a `base` property? Something like `if (property_exists($uri, 'base')) {` should work. My concern is that we could get vanilla Uri objects and hit runtime errors.

File: tests/bootstrap.php [hunk 2]
```diff
@@ -57,3 +57,5 @@
 Plugin::load('Auth', [
     'path' => dirname(dirname(__FILE__)) . DS,
 ]);
+
+$_SERVER['PHP_SELF'] = '/';
```
Comment (markstory) [inline tests/bootstrap.php]: This is reasonable to me. In a web context `PHP_SELF` would not be a file path to vendor/

File: src/UrlChecker/DefaultUrlChecker.php [hunk 3]
```diff
@@ -102,6 +102,8 @@ protected function _getChecker(array $options = [])
      */
     protected function _getUrlFromRequest(UriInterface $uri, $getFullUrl = false)
     {
+        $uri = $uri->withPath($uri->base . $uri->getPath());
```
Comment (markstory) [inline src/UrlChecker/DefaultUrlChecker.php]: Would an application in a base directory have it `loginUrl` setup to have the base path?

Comment (Zuluru) [reply src/UrlChecker/DefaultUrlChecker.php]: I'm not clear on the question. Is your concern that `$uri->base` will never be non-empty here, or that the resulting URL will have the base duplicated?

Comment (markstory) [reply src/UrlChecker/DefaultUrlChecker.php]: My question is more on how loginUrl is generally used for applications in subdirectories. Is it configured to be `/base/login` or `/login`?

File: src/Middleware/AuthenticationMiddleware.php (hunk 1 above)
Comment (Zuluru) [reply src/Middleware/AuthenticationMiddleware.php]: The uri we deal with here is the one parsed by the request, which I believe will always have the base set, though it may be null. Unless we expect this may be used with something other than Cake's ServerRequestFactory?

**Comment by Zuluru** [comment]
There are also two places in `Cake\Http\Cookie\CookieCollection` (`addFromResponse` and `addToRequest`) that use `$uri->getPath()`. I don't really use cookies outside of authentication, so it's not clear to me whether those should also be updated in a similar manner.

File: src/UrlChecker/DefaultUrlChecker.php (hunk 3 above)
Comment (Zuluru) [reply src/UrlChecker/DefaultUrlChecker.php]: Just realized that your concern may be that if we make this URL include the base, but the configured one didn't have that as part of it, then they won't match. I guess my response would be that, since this already isn't working in subfolders, there shouldn't really be any code out there relying on any particular behaviour. And since the URL checker does strict comparisons without any type conversion, it has to be given string URLs in the configuration, and the sensible way of generating those strings is with `Router::url()`, which will include the base.

In short, including the base here shouldn't break existing code, and works as expected with `Router::url()`.

File: src/Middleware/AuthenticationMiddleware.php (hunk 1 above)
Comment (Zuluru) [reply src/Middleware/AuthenticationMiddleware.php]: Not sure if you're waiting on me to update the PR based on this...

Comment (markstory) [reply src/Middleware/AuthenticationMiddleware.php]: We tried to make this middleware usable in non-cakephp projects. Doing so requires sticking to the PSR7 interface methods and attributes and only conditionally using our extensions if they are available.

File: tests/bootstrap.php (hunk 2 above)
Comment (Zuluru) [reply tests/bootstrap.php]: My git-fu isn't strong enough to sort out how to resolve this conflict. :-(

Comment (Zuluru) [reply tests/bootstrap.php]: Should have some time to get back to this on the weekend, now that a big deadline on another project is past. Anyone have suggestions on what git commands I need to run to merge the other commits into my own fork, resolve the conflict, and push back into my PR?

Comment (garas) [reply tests/bootstrap.php]: Run `git merge upstream/master` (assuming this repo is named `upstream`). It will say `bootstrap.php` has conflict.

Open `bootstrap.php`, near end it will have
```
<<<<<<< master
Plugin::load('Auth', [
    'path' => dirname(dirname(__FILE__)) . DS,
]);

$_SERVER['PHP_SELF'] = '/';
=======
Plugin::getCollection()->add(new \Authentication\Plugin());
>>>>>>> master
```

Resolve conflict as 
```
Plugin::getCollection()->add(new \Authentication\Plugin());

$_SERVER['PHP_SELF'] = '/';
```

Now add updated file, commit and push to your fork as you always do (ex. `git add bootstrap.php`, `git commit`, `git push`, or use GUI tools)

Comment (Zuluru) [reply tests/bootstrap.php]: Awesome, thanks! I'm sure I tried something close to that, but obviously didn't hit the right combo. I mostly work in a "team" of one, so my git usage tends to be pretty straight-line.

Comment (garas) [reply tests/bootstrap.php]: You've resolved conflict incorrectly...

In #233, the `Plugin::load(...);` was replaced with `Plugin::getCollection()...`. However you left both of them.

Your PR adds `$_SERVER['PHP_SELF'] = '/';` after that line.

Remove `Plugin::load(...);`. After that, _Files changed_ tab above should only show the line you've added. 

Also there is no new line now after `$_SERVER['PHP_SELF'] = '/';`.

Comment (Zuluru) [reply tests/bootstrap.php]: Yep, that was a foolish oversight, skipped through too quickly specifically because it was such a small change. :-(
